src/plugins/intel_cpu/src/transformations/cpu_opset/common/pass/move_readvalue_inputs_to_subgraph.cpp

```diff
@@ -0,0 +1,141 @@
+#include "transformations/rt_info/disable_fp16_compression.hpp"
+#include "transformations/utils/gen_pattern.hpp"
+#include "transformations/utils/utils.hpp"
+#include "transformations/cpu_opset/common/op/submodel.hpp"
+
+ov::intel_cpu::MoveReadValueInputsToSubgraph::MoveReadValueInputsToSubgraph() {
+    MATCHER_SCOPE(MoveReadValueInputsToSubgraph);
+    using namespace ov::pass::pattern;
+
+    auto readvalue_pattern = pass::pattern::wrap_type<ov::op::v6::ReadValue>();
+
+    ov::matcher_pass_callback callback = [=](Matcher& m) {
+        const auto& pattern_map = m.get_pattern_value_map();
+        auto readvalue = as_type_ptr<ov::opset6::ReadValue>(pattern_map.at(readvalue_pattern).get_node_shared_ptr());
+        if (!readvalue || readvalue->get_input_size() != 1u) {
+            return false;
+        }
+
+        bool found_output = false;  // Flag: find Output node or out of max depth.
+        std::string root_name = readvalue->get_friendly_name();
+
+        NodeVector subgraph_nodes;
+        std::unordered_set<std::string> subgraph_node_names;
+        std::unordered_set<std::shared_ptr<ov::Node>> visited;  // Visited flag.
+        NodeVector inputs = {};
+        OutputVector outputs = {};
+
+        // DFS, Check if current node's final successor is only ReadValue.
+        std::function<void(std::shared_ptr<ov::Node>)> dfs = [&](std::shared_ptr<ov::Node> node) {
+            if (visited.find(node) != visited.end()) {
+                return;
+            }
+
+            // Output node.
+            if (node->get_output_target_inputs(0).size() == 0u) {
+                found_output = true;
+                return;
+            }
+
+            for (const auto& child : node->get_output_target_inputs(0)) {
+                auto son = child.get_node()->shared_from_this();
+                if (son->get_friendly_name() == root_name) {
+                    continue;
+                }
+                dfs(son);
+            }
+        };
+
+        std::function<void(std::shared_ptr<ov::Node>)> reverse_dfs = [&](std::shared_ptr<ov::Node> node) {
+            if (ov::op::util::is_parameter(node)) {
+                inputs.emplace_back(node);
+                return;
+            }
+
+            // Check whether current node have same successor[root_node_name].
+            found_output = false;
+            dfs(node);
+            visited.insert(node);
+            if (found_output) {
+                inputs.emplace_back(node);
+                return;
+            }
+
+            for (size_t i = 0; i < node->get_input_size(); i++) {
+                reverse_dfs(node->get_input_node_shared_ptr(i));
+            }
+
+            // Cache to subgraph_nodes
+            subgraph_nodes.emplace_back(node);
+            subgraph_node_names.insert(node->get_friendly_name());
+        };
+
+        // Reverse DFS ReadValue, find all suitable nodes and move them to subgraph_nodes.
+        reverse_dfs(readvalue->get_input_node_shared_ptr(0));
```

Comment:
@maxnick About time and space complexity, here is my analysis:
Our graph is sparse and directed.
1: reverse_dfs only access top node of ReadValue.
**Time complexity** should be O(V+E), V is Vertex, E is edge, just ignore edge, final time complexity is O(n).
**Space complexity**: because only save node to subgraph_nodes and node's name once, it should be O(2n).

2: For each node in the reverse DFS, do DFS.
**Time complexity:** because we flagged all visited node, if a node accessed, it will be skipped. And it will exit all DFS if accessed a node which connect to Output and bypass ReadValue. So it also be O(N).
**Space complexity:** For accessed node, save to Visited once, so space complexity only should be O(n).

3: Merge 2 searches:
**Time complexity** is O(2n), final is O(n)
**Space complexity** is O(3n), final is O(n)